File: class/Locate.php
<?php

require_once 'Qwik.php';
require_once 'Defend.php';
require_once 'Service.php';


class Locate extends Qwik {
    
    static $geoplace;
    static $geodetails;    
    static $geotimezone;    
    static $geocode;

    // https://stackoverflow.com/questions/693691/how-to-initialize-static-variables
    static function initStatic(){
        self::$geoplace = new Service("geoplace");
        self::$geodetails = new Service("geodetails");
        self::$geotimezone = new Service("geotimezone");
        self::$geocode = new Service("geocode");
    }


    static function geo($param, $key, $url){
        $result = NULL;
        $param['key'] = $key;
        try{
            $query = http_build_query($param);
            $json = json_decode(Defend::json("$url?$query"), TRUE);
            $status = isset($json['status']) ? $json['status'] : "json reply mising status";
            if($status === 'OK' || $status === 'ZERO_RESULTS'){
                $result = $json;
            } else {
                throw new RuntimeException($status);
            }
        } catch (RuntimeException $e){
            $msg = $e->getMessage();
            self::logMsg("Google geocoding: $msg\n$url?$query\n$result");
        }
        return $result;
    }


    static function geoplace($text, $country){
        return self::geo(
            array('input'=>$text, 'components'=>"country:$country"),
            self::$geoplace->key('private'),
            self::$geoplace->url("json")
        );
    }


    static function geodetails($placeid){
        $geo = self::geo(
            array('place_id'=>$placeid),
            self::$geodetails->key('private'),
            self::$geodetails->url("json")
        );
        return isset($geo['result']) ? $geo['result'] : NULL;
    }


    static function geotime($lat, $lng){
        $location = "$lat,$lng";
        return self::geo(
            array('location'=>$location, 'timestamp'=>time()),
            self::$geotimezone->key('private'),
            self::$geotimezone->url("json")
        );
    }


    static function geocode($address, $country){
        $geo = self::geo(
            array('address'=>"$address", 'components'=>"country:$country"),
            self::$geocode->key('private'),
            self::$geocode->url("json")
        );
        return isset($geo['result']) ? $geo['result'] : NULL;
    }


    static function revgeocode($lat, $lng){
      $type = "country|administrative_area_level_1|locality";
      $geo = self::geo(
        array('latlng'=>"$lat,$lng", 'result_type'=>$type),
        self::$geocode->key('private'),
        self::$geocode->url("json")
      );
      return isset($geo['results']) ? $geo['results'] : NULL;
    }



    static function getPlace($description, $country){
        $placeid = NULL;
        $geoplace = self::geoplace($description, $country);
        if(isset($geoplace['predictions'][0])){
            $place = $geoplace['predictions'][0];
            $placeid = (string) $place['place_id'];
        }
        return $placeid;
    }


    static function getDetails($placeid){
        $details = NULL;
        $result = self::geodetails($placeid);
        if(isset($result)){
          $details = array();
          $details['placeid'] = $placeid;

          $details['name'] = isset($result['name']) ? (string) $result['name'] : '';
          $details['address'] = isset($results['formatted_address']) ? (string) $results['formatted_address'] : '';

          if (isset($result['geometry']['location'])){
            $location = $result['geometry']['location'];
            $lat = (string) $location['lat'];
            $lng = (string) $location['lng'];
            $details['lat'] = $lat;
            $details['lng'] = $lng;
            $details['tz'] = self::getTimezone($lat, $lng);
          }

          if (isset($result['address_components'])){
            $address_components = $result['address_components'];
            foreach($address_components as $addr){
              $types = isset($addr['types']) ? $addr['types'] : array() ;
              foreach($types as $type){
                switch ((string) $type){
                  case 'country':
                    $details['country'] = (string) $addr['long_name'];
                    $details['country_iso'] = (string) $addr['short_name'];
                    break;
                  case 'administrative_area_level_1':
                    $details['admin1'] = (string) $addr['long_name'];
                    $details['admin1_code'] = (string) $addr['short_name'];
                    break;
                  case 'street_number':
                    $details['str-num'] = (string) $addr['long_name'];
                    break;
                  case 'route':
                    $details['route'] = (string) $addr['long_name'];
                    break;
                  case 'locality':
                    $details['locality'] = (string) $addr['long_name'];
                    break;
                  default:
                }
              }
            }
          }

          $details['phone'] = isset($result['phone']) ? (string) $result['phone'] : '';
          $details['phone'] = isset($result['international_phone_number']) ? (string) $result['international_phone_number'] : '';
          $details['url'] = isset($result['website']) ? (string) $result['website'] : '';
        }
        return $details;
    }


    static function getTimezone($lat, $lng){
        $geotime = self::geotime($lat, $lng);
        return isset($geotime['timeZoneId']) ? (string) $geotime['timeZoneId'] : '';
    }


    static function parseAddress($address, $country=NULL){
        $parsed = FALSE;
        $placeid = self::getPlace($address, $country);
        if (isset($placeid)){
            $parsed = self::getDetails($placeid);
        }
        return $parsed;
    }


    static function guessTimezone($location, $admin1, $country){
        $tz = NULL;
        $placeid = self::getPlace("$location, $admin1", $country);
        if(isset($placeid)){
            $details = self::geodetails($placeid);
            if(isset($details['geometry']['location'])){
                $loc = $details['geometry']['location'];
                $tz = self::getTimezone($loc['lat'], $loc['lng']);
            }
        }
        return $tz;
    }
    

    static function geolocate($key){
        global $geo;
        if(!isset($geo)){
            $geo = unserialize(file_get_contents('http://www.geoplugin.net/php.gp?ip='.$_SERVER['REMOTE_ADDR']));
            if(!isset($geo)){ return NULL; }   // geoplugin.net is offline
        }
        if(is_array($key)){
            $result = array();
            foreach($key as $k){
                $geoKey = "geoplugin_$k";
                $result[$k] = isset($geo[$geoKey]) ? $geo[$geoKey] : NULL ;
            }
            return $result;
        } else {
            $geoKey = "geoplugin_$key";
            return isset($geo[$geoKey]) ? $geo[$geoKey] : NULL ;
        }
    }


    public function __construct(){
        parent::__construct();
    }


    static function geoGuess($input){
        $result = "{}";
    	$param = array();
        $param['input'] = $input;
        $param['key'] = self::$geoplace->key('private');
        $url = self::$geoplace->url("json");
        try{
            $query = http_build_query($param);
            $json = Defend::json("$url?$query");
            $decoded = json_decode($json, TRUE);
            $status = (string) $decoded["status"];
            if($status === 'OK' || $status === 'ZERO_RESULTS'){
                $result = $json;
            } else {
                throw new RuntimeException($status);
            }
        } catch (RuntimeException $e){
            $msg = $e->getMessage();
            self::logMsg("Google geocoding: $msg\n$url?$query\n$msg");
        }
        return $result;
    }


  static function getAddress($lat, $lng){
    $address = [];
    $revgeocode = self::revgeocode($lat, $lng);
    if(isset($revgeocode[0]['address_components'])){
      $components = $revgeocode[0]['address_components'];
      foreach($components as $component){
        $types = isset($component['types']) ? $component['types'] : array();
        foreach($types as $type){
          switch ((string) $type){
            case 'country':
              $address['country']  = (string) $component['short_name'];
              break ;
            case 'administrative_area_level_1':
              $name = (string) $component['short_name'];
              $name = empty($name) ? (string) $component['long_name'] : $name ;
              $name = empty($name) ? ' ' : $name ;
              $address['admin1']   = $name;
              break ;
            case 'locality':
              $address['locality'] = (string) $component['short_name'];
              break ;
          }
        }
      }
    }
    return $address;
  }


  static function getGeometry($country, $admin1, $locality){
    $geometry = NULL;
    if(!empty($admin1) && !empty($locality)){
      $input = "$locality, $admin1";
    } elseif (!empty($admin1)){
      $input = $admin1;
    } elseif (!empty($locality)){
      $input = $locality;
    } elseif (isset(Qwik::countries()[$country])) {
      $input = Qwik::countries()[$country];
    } else {
      return;
    } 
    $placeid = self::getPlace($input, $country);
    if(isset($placeid)){
      $details = self::geodetails($placeid);
      if (isset($details['geometry'])){
        $geometry = $details['geometry'];
      }
    }
    return $geometry;
  }

}


Locate::initStatic();

?>

File: service/locate.py
import json, logging, requests, time, urllib.parse
from service.models import Service
from qwikgame.constants import COUNTRIES
from qwikgame.hourbits import Hours24x7, WEEK_ALL

logger = logging.getLogger(__file__)

GEOPLACE = 'geoplace'
GEODETAILS = 'geodetails'
GEOTIMEZONE = 'geotimezone'
GEOCODE = 'geocode'
GEOPLUGIN_CONTEXT = {"http": {"timeout": 1}}

class Locate:

    @staticmethod   
    def geo(param, key, url):
        result = None
        param['key'] = key
        try:
            response = requests.post(url, params=param)
            response.raise_for_status()
            result = response.json()
        except:
            logger.exception("Google geocoding: {}?{}\n{}".format(url, param, response))
        return result


    @staticmethod
    def geoplace(text, country=None, region=None):
        geoplace = Service.objects.get(pk=GEOPLACE)
        param = {'input': text}
        if country:
            param['components'] = f'country:{country}'
        if region:
            param['components'] = f'country:{region.country}'
            param['locationbias'] = f'rectangle:{region.south},{region.west}|{region.north},{region.east}'
        return Locate.geo(
            param,
            geoplace.key,
            geoplace.url
        )


    @staticmethod
    def geodetails(placeid):
        geodetails = Service.objects.get(pk=GEODETAILS)
        geo = Locate.geo(
            {'place_id': placeid},
            geodetails.key,
            geodetails.url
        )
        if 'result' in geo:
            return geo['result']
        logger.warn('no result from geodetails: {geo}')
        return None


    @staticmethod
    def geotime(lat, lng):
        geotimezone = Service.objects.get(pk=GEOTIMEZONE)
        location = f"{lat},{lng}"
        return Locate.geo(
            {'location': location, 'timestamp': time.time()},
            geotimezone.key,
            geotimezone.url
        )


    @staticmethod
    def geocode(address, country):
        geocode = Service.objects.get(pk=GEOCODE)
        geo = Locate.geo(
            {'address': address, 'components': f"country:{country}"},
            geocode.key,
            geocode.url
        )
        if 'result' in geo:
            return geo['result']
        logger.warn('no result from geocode: {geo}')
        return None


    @staticmethod
    def revgeocode(lat, lng):
        geocode = Service.objects.get(pk=GEOCODE)
        type_ = "country|administrative_area_level_1|locality"
        geo = Locate.geo(
            {'pos': f"{lat},{lng}", 'result_type': type_},
            geocode.key,
            geocode.url
        )
        if 'result' in geo:
            return geo['result']
        logger.warn('no result from revgeocode: {geo}')
        return None


    @staticmethod
    def get_place(description, country):
        placeid = None
        geoplace = Locate.geoplace(description, country)
        if 'predictions' in geoplace and len(geoplace['predictions']) > 0:
            place = geoplace['predictions'][0]
            placeid = str(place['place_id'])
        return placeid


    @staticmethod
    def get_places(description, region=None):
        places = {}
        geoplace = Locate.geoplace(description, region=region)
        predictions = geoplace.get('predictions', None)
        logger.info(predictions)
        return { str(place['place_id']): str(place['description']) for place in predictions}


    @staticmethod
    def get_details(placeid):
        details = None
        result = Locate.geodetails(placeid)
        if result is not None:
            details = {}
            details['placeid'] = placeid
            details['name'] = result.get('name', '')
            details['address'] = result.get('formatted_address', '')
            details['url'] = result.get('website', '')
            details['phone'] = result.get('international_phone_number', '').replace(' ','')
            editorial_summary = result.get('editorial_summary', None)
            if editorial_summary:
                details['note'] = editorial_summary.get('overview' '')
            geometry = result.get('geometry', None)
            if geometry:
                location = geometry.get('location', None)
                if location:
                    lat = location.get('lat', 0)
                    lng = location.get('lng', 0)
                    details['lat'] = lat
                    details['lng'] = lng
                    details['tz'] = Locate.get_timezone(lat, lng)  # Replace with the actual class name
            for comp in result.get('address_components', []):
                types = comp.get('types', None)
                if types:
                    if 'country' in types:
                        details['country'] = comp.get('short_name', '')
                        # details['country_long'] = comp.get('long_name', '')
                    elif 'administrative_area_level_1' in types:
                        details['admin1'] = comp.get('short_name', '')
                        # details['admin1_long'] = comp.get('long_name', '')
                    elif 'administrative_area_level_2' in types:
                        details['suburb'] = comp.get('short_name', '')
                    elif 'locality' in types: 
                        details['locality'] = comp.get('short_name', '')
                        # details['locality_long'] = comp.get('short_name', '')
                    elif 'route' in types:
                        details['route'] = comp.get('short_name', '')
                    elif 'street_number' in types:
                        details['str_num'] = comp.get('short_name', '')

            opening_hours = result.get('opening_hours', None)
            if opening_hours:
                hours = Hours24x7()
                for period in opening_hours.get('periods', []):
                    open = period.get('open')
                    close = period.get('close')
                    if open and close:
                        try:
                            open_day = int(open.get('day'))
                            open_hour = int(open.get('time'))
                            close_day = int(close.get('day'))
                            close_hour = int(close.get('time'))
                            first_hour = (open_hour // 100) if ((open_hour % 100) == 0) else ((open_hour // 100) + 1)
                            last_hour = (close_hour // 100)
                            hours.set_period(open_day, first_hour, close_day, last_hour)
                        except:
                            logger.exception('Invalid opening hours period {period}')
                details['hours'] = hours.as_bytes()
            else:
                details['hours'] = WEEK_ALL
                logger.warn(f'opening_hours unavailable ({placeid}) - default to 24x7')
        return details


    @staticmethod
    def get_timezone(lat, lng):
        geotime = Locate.geotime(lat, lng)
        return str(geotime['timeZoneId']) if 'timeZoneId' in geotime else ''


    @staticmethod
    def parse_address(address, country=None):
        parsed = False
        placeid = Locate.get_place(address, country)
        if placeid is not None:
            parsed = Locate.get_details(placeid)
        return parsed


    @staticmethod
    def guess_timezone(location, admin1, country):
        tz = None
        placeid = Locate.get_place(f"{location}, {admin1}", country)
        if placeid is not None:
            details = Locate.geodetails(placeid)
            if 'geometry' in details and 'location' in details['geometry']:
                loc = details['geometry']['location']
                tz = Locate.get_timezone(loc['lat'], loc['lng'])
        return tz


    @staticmethod
    def geolocate(key):
        global geo
        if 'geo' not in globals():
            if 'REMOTE_ADDR' in os.environ:
                remote_add = os.environ['REMOTE_ADDR']
                url = f"http://www.geoplugin.net/php.gp?ip={remote_add}"
                response = requests.get(url)
                geo = json.loads(response.text)
                if geo is None:
                    return None  # geoplugin.net is offline
        if isinstance(key, list):
            result = {}
            for k in key:
                geo_key = f"geoplugin_{k}"
                result[k] = geo.get(geo_key, None)
            return result
        else:
            geo_key = f"geoplugin_{key}"
            return geo.get(geo_key, None)

    def __init__(self):
        super().__init__()


    @staticmethod
    def geo_guess(input):
        result = "{}"
        param = {
            'input': input,
            'key': Locate.geoplace.key('private')
        }
        url = Locate.geoplace.url("json")
        try:
            query = urllib.parse.urlencode(param)
            response = requests.get(f"{url}?{query}")
            json_data = response.text
            decoded = json.loads(json_data)
            status = str(decoded["status"])
            if status in ['OK', 'ZERO_RESULTS']:
                result = json_data
            else:
                raise RuntimeError(status)
        except RuntimeError as e:
            msg = str(e)
            logger.warn(f"Google geocoding: {msg}\n{url}?{query}\n{msg}")
        return result

    @staticmethod
    def get_address(lat, lng):
        address = {}
        revgeocode = Locate.revgeocode(lat, lng)
        if revgeocode and 'address_components' in revgeocode[0]:
            components = revgeocode[0]['address_components']
            for component in components:
                types = component.get('types', [])
                for type in types:
                    if type == 'country':
                        address['country'] = str(component['short_name'])
                    elif type == 'administrative_area_level_1':
                        name = str(component['short_name']) or str(component['long_name']) or ' '
                        address['admin1'] = name
                    elif type == 'locality':
                        address['locality'] = str(component['short_name'])
        return address


    @staticmethod
    def get_geometry(country, admin1, locality):
        result = None
        if admin1 and locality:
            input_str = f"{locality}, {admin1}"
        elif admin1:
            input_str = admin1
        elif locality:
            input_str = locality
        elif country in COUNTRIES:
            input_str = COUNTRIES[country]
        else:
            logger.warn(f"Locate::getGeometry({country}, {admin1}, {locality}) insufficient parameters")
            return
        placeid = Locate.get_place(input_str, country)
        if placeid:
            details = Locate.geodetails(placeid)
            if details:
                result = details.get('geometry', {})
                result['placeid'] = details.get('place_id', placeid)
                result['names'] = {}
                tipes = ['country', 'admin1', 'locality']
                for comp in details.get('address_components', []):
                    tipes = comp.get('types', [])
                    if 'country' in tipes:
                        result['names']['country'] = comp['long_name'] or comp['short_name'] or country
                    if 'administrative_area_level_1' in tipes:
                        result['names']['admin1'] = comp['long_name'] or comp['short_name'] or admin1
                    if 'locality' in tipes:
                        result['names']['locality'] = comp['long_name'] or comp['short_name'] or locality
        return result
